File: data/country-tiers.ts
import type { Difficulty } from "@/lib/game-engine/types";

/**
 * Country tier system for difficulty-based filtering.
 *
 * Tier 1 — Well-known countries: large, distinctive shape, globally recognized (pop >20M or very famous)
 * Tier 2 — Moderately known: regional significance, medium-sized, recognizable to geography enthusiasts
 * Tier 3 — Obscure/micro: small, micro-states, easily confused, or rarely referenced
 *
 * Used by games to scale the country pool:
 *   Easy   → Tier 1 only
 *   Medium → Tier 1 + 2
 *   Hard   → All tiers (1 + 2 + 3)
 */
export type CountryTier = 1 | 2 | 3;

export const countryTiers: Record<string, CountryTier> = {
  // ─── EUROPE ───
  AL: 2, // Albania
  AD: 3, // Andorra
  AT: 2, // Austria
  BY: 2, // Belarus
  BE: 2, // Belgium
  BA: 3, // Bosnia and Herzegovina
  BG: 2, // Bulgaria
  HR: 2, // Croatia
  CZ: 2, // Czech Republic
  DK: 2, // Denmark
  EE: 3, // Estonia
  FI: 2, // Finland
  FR: 1, // France
  DE: 1, // Germany
  GR: 1, // Greece
  HU: 2, // Hungary
  IS: 2, // Iceland
  IE: 2, // Ireland
  IT: 1, // Italy
  XK: 3, // Kosovo
  LV: 3, // Latvia
  LI: 3, // Liechtenstein
  LT: 3, // Lithuania
  LU: 3, // Luxembourg
  MK: 3, // North Macedonia
  MT: 3, // Malta
  MD: 3, // Moldova
  MC: 3, // Monaco
  ME: 3, // Montenegro
  NL: 2, // Netherlands
  NO: 1, // Norway
  PL: 1, // Poland
  PT: 2, // Portugal
  RO: 2, // Romania
  RU: 1, // Russia
  SM: 3, // San Marino
  RS: 2, // Serbia
  SK: 3, // Slovakia
  SI: 3, // Slovenia
  ES: 1, // Spain
  SE: 1, // Sweden
  CH: 2, // Switzerland
  UA: 1, // Ukraine
  GB: 1, // United Kingdom
  VA: 3, // Vatican City

  // ─── ASIA ───
  AF: 2, // Afghanistan
  AM: 3, // Armenia
  AZ: 3, // Azerbaijan
  BH: 3, // Bahrain
  BD: 2, // Bangladesh
  BT: 3, // Bhutan
  BN: 3, // Brunei
  KH: 2, // Cambodia
  CN: 1, // China
  CY: 3, // Cyprus
  GE: 3, // Georgia
  IN: 1, // India
  ID: 1, // Indonesia
  IR: 1, // Iran
  IQ: 1, // Iraq
  IL: 2, // Israel
  JP: 1, // Japan
  JO: 2, // Jordan
  KZ: 2, // Kazakhstan
  KW: 3, // Kuwait
  KG: 3, // Kyrgyzstan
  LA: 3, // Laos
  LB: 3, // Lebanon
  MY: 2, // Malaysia
  MV: 3, // Maldives
  MN: 2, // Mongolia
  MM: 2, // Myanmar
  NP: 2, // Nepal
  KP: 2, // North Korea
  OM: 3, // Oman
  PK: 1, // Pakistan
  PS: 3, // Palestine
  PH: 2, // Philippines
  QA: 3, // Qatar
  SA: 1, // Saudi Arabia
  SG: 3, // Singapore
  KR: 1, // South Korea
  LK: 2, // Sri Lanka
  SY: 2, // Syria
  TW: 2, // Taiwan
  TJ: 3, // Tajikistan
  TH: 1, // Thailand
  TL: 3, // Timor-Leste
  TR: 1, // Turkey
  TM: 3, // Turkmenistan
  AE: 2, // United Arab Emirates
  UZ: 3, // Uzbekistan
  VN: 1, // Vietnam
  YE: 2, // Yemen

  // ─── AFRICA ───
  DZ: 1, // Algeria
  AO: 2, // Angola
  BJ: 3, // Benin
  BW: 3, // Botswana
  BF: 3, // Burkina Faso
  BI: 3, // Burundi
  CV: 3, // Cape Verde
  CM: 2, // Cameroon
  CF: 3, // Central African Republic
  TD: 2, // Chad
  KM: 3, // Comoros
  CG: 3, // Republic of the Congo
  CD: 1, // Democratic Republic of the Congo
  CI: 2, // Ivory Coast
  DJ: 3, // Djibouti
  EG: 1, // Egypt
  GQ: 3, // Equatorial Guinea
  ER: 3, // Eritrea
  SZ: 3, // Eswatini
  ET: 1, // Ethiopia
  GA: 3, // Gabon
  GM: 3, // Gambia
  GH: 2, // Ghana
  GN: 3, // Guinea
  GW: 3, // Guinea-Bissau
  KE: 1, // Kenya
  LS: 3, // Lesotho
  LR: 3, // Liberia
  LY: 2, // Libya
  MG: 2, // Madagascar
  MW: 3, // Malawi
  ML: 2, // Mali
  MR: 3, // Mauritania
  MU: 3, // Mauritius
  MA: 2, // Morocco
  MZ: 2, // Mozambique
  NA: 2, // Namibia
  NE: 2, // Niger
  NG: 1, // Nigeria
  RW: 3, // Rwanda
  ST: 3, // Sao Tome and Principe
  SN: 2, // Senegal
  SC: 3, // Seychelles
  SL: 3, // Sierra Leone
  SO: 2, // Somalia
  ZA: 1, // South Africa
  SS: 3, // South Sudan
  SD: 2, // Sudan
  TZ: 1, // Tanzania
  TG: 3, // Togo
  TN: 2, // Tunisia
  UG: 2, // Uganda
  ZM: 2, // Zambia
  ZW: 2, // Zimbabwe

  // ─── NORTH AMERICA ───
  AG: 3, // Antigua and Barbuda
  BS: 3, // Bahamas
  BB: 3, // Barbados
  BZ: 3, // Belize
  CA: 1, // Canada
  CR: 2, // Costa Rica
  CU: 1, // Cuba
  DM: 3, // Dominica
  DO: 2, // Dominican Republic
  SV: 3, // El Salvador
  GD: 3, // Grenada
  GT: 2, // Guatemala
  HT: 2, // Haiti
  HN: 3, // Honduras
  JM: 2, // Jamaica
  MX: 1, // Mexico
  NI: 3, // Nicaragua
  PA: 2, // Panama
  KN: 3, // Saint Kitts and Nevis
  LC: 3, // Saint Lucia
  VC: 3, // Saint Vincent and the Grenadines
  TT: 3, // Trinidad and Tobago
  US: 1, // United States

  // ─── SOUTH AMERICA ───
  AR: 1, // Argentina
  BO: 2, // Bolivia
  BR: 1, // Brazil
  CL: 1, // Chile
  CO: 1, // Colombia
  EC: 2, // Ecuador
  GY: 3, // Guyana
  PY: 2, // Paraguay
  PE: 1, // Peru
  SR: 3, // Suriname
  UY: 2, // Uruguay
  VE: 1, // Venezuela
  GF: 3, // French Guiana

  // ─── OCEANIA ───
  AU: 1, // Australia
  FJ: 3, // Fiji
  NZ: 1, // New Zealand
  PG: 2, // Papua New Guinea
  WS: 3, // Samoa
  SB: 3, // Solomon Islands
  TO: 3, // Tonga
  VU: 3, // Vanuatu
};

/** Get the tier for a country code (defaults to tier 3 if not found) */
export function getCountryTier(code: string): CountryTier {
  return countryTiers[code] ?? 3;
}

/** Get the maximum tier allowed for a given difficulty */
export function getMaxTierForDifficulty(difficulty: Difficulty): CountryTier {
  switch (difficulty) {
    case "easy":
      return 1;
    case "medium":
      return 2;
    case "hard":
      return 3;
  }
}

File: geoplay-mobile/lib/core/data/country_tiers.dart
import 'countries.dart';

/// Country tier system for difficulty-based filtering.
/// Ported from data/country-tiers.ts.
///
/// Tier 1 — Well-known: large, distinctive, globally recognized
/// Tier 2 — Moderately known: regional significance, medium-sized
/// Tier 3 — Obscure/micro: small, micro-states, easily confused
enum CountryTier { tier1, tier2, tier3 }

const Map<String, CountryTier> countryTiers = {
  // ─── EUROPE ───
  'AL': CountryTier.tier2, 'AD': CountryTier.tier3, 'AT': CountryTier.tier2,
  'BY': CountryTier.tier2, 'BE': CountryTier.tier2, 'BA': CountryTier.tier3,
  'BG': CountryTier.tier2, 'HR': CountryTier.tier2, 'CZ': CountryTier.tier2,
  'DK': CountryTier.tier2, 'EE': CountryTier.tier3, 'FI': CountryTier.tier2,
  'FR': CountryTier.tier1, 'DE': CountryTier.tier1, 'GR': CountryTier.tier1,
  'HU': CountryTier.tier2, 'IS': CountryTier.tier2, 'IE': CountryTier.tier2,
  'IT': CountryTier.tier1, 'XK': CountryTier.tier3, 'LV': CountryTier.tier3,
  'LI': CountryTier.tier3, 'LT': CountryTier.tier3, 'LU': CountryTier.tier3,
  'MK': CountryTier.tier3, 'MT': CountryTier.tier3, 'MD': CountryTier.tier3,
  'MC': CountryTier.tier3, 'ME': CountryTier.tier3, 'NL': CountryTier.tier2,
  'NO': CountryTier.tier1, 'PL': CountryTier.tier1, 'PT': CountryTier.tier2,
  'RO': CountryTier.tier2, 'RU': CountryTier.tier1, 'SM': CountryTier.tier3,
  'RS': CountryTier.tier2, 'SK': CountryTier.tier3, 'SI': CountryTier.tier3,
  'ES': CountryTier.tier1, 'SE': CountryTier.tier1, 'CH': CountryTier.tier2,
  'UA': CountryTier.tier1, 'GB': CountryTier.tier1, 'VA': CountryTier.tier3,

  // ─── ASIA ───
  'AF': CountryTier.tier2, 'AM': CountryTier.tier3, 'AZ': CountryTier.tier3,
  'BH': CountryTier.tier3, 'BD': CountryTier.tier2, 'BT': CountryTier.tier3,
  'BN': CountryTier.tier3, 'KH': CountryTier.tier2, 'CN': CountryTier.tier1,
  'CY': CountryTier.tier3, 'GE': CountryTier.tier3, 'IN': CountryTier.tier1,
  'ID': CountryTier.tier1, 'IR': CountryTier.tier1, 'IQ': CountryTier.tier1,
  'IL': CountryTier.tier2, 'JP': CountryTier.tier1, 'JO': CountryTier.tier2,
  'KZ': CountryTier.tier2, 'KW': CountryTier.tier3, 'KG': CountryTier.tier3,
  'LA': CountryTier.tier3, 'LB': CountryTier.tier3, 'MY': CountryTier.tier2,
  'MV': CountryTier.tier3, 'MN': CountryTier.tier2, 'MM': CountryTier.tier2,
  'NP': CountryTier.tier2, 'KP': CountryTier.tier2, 'OM': CountryTier.tier3,
  'PK': CountryTier.tier1, 'PS': CountryTier.tier3, 'PH': CountryTier.tier2,
  'QA': CountryTier.tier3, 'SA': CountryTier.tier1, 'SG': CountryTier.tier3,
  'KR': CountryTier.tier1, 'LK': CountryTier.tier2, 'SY': CountryTier.tier2,
  'TW': CountryTier.tier2, 'TJ': CountryTier.tier3, 'TH': CountryTier.tier1,
  'TL': CountryTier.tier3, 'TR': CountryTier.tier1, 'TM': CountryTier.tier3,
  'AE': CountryTier.tier2, 'UZ': CountryTier.tier3, 'VN': CountryTier.tier1,
  'YE': CountryTier.tier2,

  // ─── AFRICA ───
  'DZ': CountryTier.tier1, 'AO': CountryTier.tier2, 'BJ': CountryTier.tier3,
  'BW': CountryTier.tier3, 'BF': CountryTier.tier3, 'BI': CountryTier.tier3,
  'CV': CountryTier.tier3, 'CM': CountryTier.tier2, 'CF': CountryTier.tier3,
  'TD': CountryTier.tier2, 'KM': CountryTier.tier3, 'CG': CountryTier.tier3,
  'CD': CountryTier.tier1, 'CI': CountryTier.tier2, 'DJ': CountryTier.tier3,
  'EG': CountryTier.tier1, 'GQ': CountryTier.tier3, 'ER': CountryTier.tier3,
  'SZ': CountryTier.tier3, 'ET': CountryTier.tier1, 'GA': CountryTier.tier3,
  'GM': CountryTier.tier3, 'GH': CountryTier.tier2, 'GN': CountryTier.tier3,
  'GW': CountryTier.tier3, 'KE': CountryTier.tier1, 'LS': CountryTier.tier3,
  'LR': CountryTier.tier3, 'LY': CountryTier.tier2, 'MG': CountryTier.tier2,
  'MW': CountryTier.tier3, 'ML': CountryTier.tier2, 'MR': CountryTier.tier3,
  'MU': CountryTier.tier3, 'MA': CountryTier.tier2, 'MZ': CountryTier.tier2,
  'NA': CountryTier.tier2, 'NE': CountryTier.tier2, 'NG': CountryTier.tier1,
  'RW': CountryTier.tier3, 'ST': CountryTier.tier3, 'SN': CountryTier.tier2,
  'SC': CountryTier.tier3, 'SL': CountryTier.tier3, 'SO': CountryTier.tier2,
  'ZA': CountryTier.tier1, 'SS': CountryTier.tier3, 'SD': CountryTier.tier2,
  'TZ': CountryTier.tier1, 'TG': CountryTier.tier3, 'TN': CountryTier.tier2,
  'UG': CountryTier.tier2, 'ZM': CountryTier.tier2, 'ZW': CountryTier.tier2,

  // ─── NORTH AMERICA ───
  'AG': CountryTier.tier3, 'BS': CountryTier.tier3, 'BB': CountryTier.tier3,
  'BZ': CountryTier.tier3, 'CA': CountryTier.tier1, 'CR': CountryTier.tier2,
  'CU': CountryTier.tier1, 'DM': CountryTier.tier3, 'DO': CountryTier.tier2,
  'SV': CountryTier.tier3, 'GD': CountryTier.tier3, 'GT': CountryTier.tier2,
  'HT': CountryTier.tier2, 'HN': CountryTier.tier3, 'JM': CountryTier.tier2,
  'MX': CountryTier.tier1, 'NI': CountryTier.tier3, 'PA': CountryTier.tier2,
  'KN': CountryTier.tier3, 'LC': CountryTier.tier3, 'VC': CountryTier.tier3,
  'TT': CountryTier.tier3, 'US': CountryTier.tier1,

  // ─── SOUTH AMERICA ───
  'AR': CountryTier.tier1, 'BO': CountryTier.tier2, 'BR': CountryTier.tier1,
  'CL': CountryTier.tier1, 'CO': CountryTier.tier1, 'EC': CountryTier.tier2,
  'GY': CountryTier.tier3, 'PY': CountryTier.tier2, 'PE': CountryTier.tier1,
  'SR': CountryTier.tier3, 'UY': CountryTier.tier2, 'VE': CountryTier.tier1,
  'GF': CountryTier.tier3,

  // ─── OCEANIA ───
  'AU': CountryTier.tier1, 'FJ': CountryTier.tier3, 'NZ': CountryTier.tier1,
  'PG': CountryTier.tier2, 'WS': CountryTier.tier3, 'SB': CountryTier.tier3,
  'TO': CountryTier.tier3, 'VU': CountryTier.tier3,
};

/// Get the tier for a country code (defaults to tier 3 if not found).
CountryTier getCountryTier(String code) {
  return countryTiers[code] ?? CountryTier.tier3;
}

/// Get the maximum tier allowed for a given difficulty.
CountryTier getMaxTierForDifficulty(String difficulty) {
  switch (difficulty) {
    case 'easy':
      return CountryTier.tier1;
    case 'medium':
      return CountryTier.tier2;
    case 'hard':
    default:
      return CountryTier.tier3;
  }
}

/// Generic country pool filter — used by ALL game engines.
/// Returns ISO codes of countries within the specified difficulty tier
/// and continent filter.
List<String> getCountryPool(String difficulty, String continent) {
  final maxTier = getMaxTierForDifficulty(difficulty);
  return countries
      .where((c) => getCountryTier(c.code).index <= maxTier.index)
      .where((c) => continent == 'all' || c.continent == continent)
      .map((c) => c.code)
      .toList();
}
